04-onion-routing.md

```diff
@@ -269,7 +272,81 @@ The writer:
 The reader:
   - MUST return an error if `amt_to_forward` or `outgoing_cltv_value` are not present.
 
-The requirements for the contents of these fields are specified [above](#legacy-hop_data-payload-format).
+The requirements for the contents of these fields are specified [above](#legacy-hop_data-payload-format)
+and [below](#basic-multi-part-payments).
+
+### Basic Multi-Part Payments
+
+If the final node receives an onion packet with the `basic_mpp` field,
+then the payment MAY be a "base" atomic multipath payment.  Such
+"base" atomic multipath payments will use the same `payment_hash` for
+all paths.
+
+The `amt_to_forward` value will be the amount for this partial payment
+only.  The `basic_mpp` flag is a promise by the ultimate sender that
+the rest of the payment will follow in succeeding HTLCs; we call these
+HTLCs, which have the same preimage, an "HTLC set".
+
+#### Requirements
+
+The writer:
+  - MUST NOT include `basic_mpp` for any non-final node.
+  - if `basic_mpp` feature was not set in the invoice:
+    - MUST NOT include `basic_mpp` for the final node.
+  - otherwise:
+    - MAY include `basic_mpp` for the final node.
+    - if it does include `basic_mpp`:
+      - MAY send more than one HTLC to pay the invoice.
+      - MUST use the same `payment_preimage` on all HTLCs in the set.
+      - MUST set `total_msat` to the amount it wishes to pay.
+      - SHOULD send all payments at approximately the same time.
+      - SHOULD try to use diverse paths to the recipient for each HTLC.
+      - SHOULD retry and/or re-divide HTLCs which fail.
+      - if the invoice specifies an `amount`:
+          - MUST set `total_msat` to at least that `amount`, and less
+            than or equal to twice `amount`.
+      - MUST ensure that the total `amount_msat` of the HTLC set which arrive at the payee
+        is greater or equal to `total_msat`.
+      - MUST ensure that no (incomplete) non-failed subset of the HTLC set adds
+        to a total `amount_msat` greater or equal to `total_msat`.
+
+The reader:
+  - if `basic_mpp` is present:
+    - MUST fail the HTLC if it would otherwise fail a single HTLC of value
+      `total_msat`, `payment_hash` or `ctlv_expiry`.
+    - otherwise:
+      - MUST add it to the HTLC set corresponding to that `payment_hash`.
+      - SHOULD fail the entire HTLC set if `total_msat` is not the same for
+        all HTLCs in the set.
```

Comment:
Does this mean that an adversary can prevent an amp payment from completion by periodically sending a very small shard with a random `total_msat`, thereby cancelling any shards that have been acquired?

Reply:
I guess those htlcs are cancelled early in the process because their `total_msat` doesn't match what is in the invoice database. How would it work in that case for zero-valued invoices? First htlc that arrives sets the amount?

Reply:
You're right, will change to ignore & fail if total_msat is different to avoid griefing.

Reply:
We must make sure that an attacker can't cancel other shards, but also that they can't block the genuine shards from being accepted. If the attacker is the first to set the tracked `total_msat` to a random value, they could possibly block the payment for the timeout duration (and repeat this thereafter).

Reply:
In that light, the proposed `randID` may actually be required not only to prevent fee stealing with 0-valued invoices, but also to keep an attacker from interfering with the multi path settle process.

Reply:
Or borrow the `stream_id` from #658 to keep the htlc sets apart. The first set that gets all its htlcs in is settled, other sets are cancelled.

Reply:
Or borrow the stream_id from #658 to keep the htlc sets apart. The first set that gets all its htlcs in is settled, other sets are cancelled.

Reply:
One thing that the receiver-generated `randID` prevents and the sender-generated not, is the attack that @t-bast described: the penultimate node can find out whether they are the penultimate node by crafting a new single hop route instead of forwarding as they should. If the newly crafted route succeeds, they know that the next node is the final destination. They still get the preimage and can pull their incoming htlc.

The attacker can generate a new sender-generated `randID` for this. But if the receiver would match on a receiver-generated `randID`, the attack wouldn't work. 

This only applies to regular, non-mpp payments I think.

Reply:
>The first set that gets all its htlcs in is settled, other sets are cancelled.

This is contradictory to being able to pay to settled invoices (to prevent leaking whether an invoice is paid or not). We could cancel all other sets, mark the invoice as settled, but if a new set comes in after that, we'd accept it again. Then we could just as well settle all sets that meet the requirements always.

Reply:
In the penultimate node attack-scenario above, I think they can also still do fee sniping on zero valued mmp invoices. The attacker crafts a new route with a new, lower `totalAmt` (single shard), and buys the preimage cheaply.

Reply:
>IIUC the attack isn't limited to the penultimate node, just easier because they have less guess on the amount, right?

Yes, sounds like it. @t-bast was there anything specific in the second-to-last case?

Reply:
Correct, anyone in the route could probe like that, but it's a lot easier for the penultimate node: it only has to test the next node in the route (which it knows because it is in the onion).

If you are the second-to-last node, you need to have the intuition of which node to probe, which is a lot harder in practice (unless you already suspect who the recipient is).

Reply:
Ok, thanks for that clarification! What about calling it "invoice secret"? (is this a better term than 'receiver-generated randid`?)

Reply:
Actually, we started calling this `payment_address` which is more descriptive in the context of #658.